homekit/StateManager.js

```diff
@@ -1,623 +1,1034 @@
-		smartModeState.highTemperatureState = {
-			on: false,
-			targetTemperature: device.state.targetTemperature,
-			temperatureUnit: device.temperatureUnit,
-			mode: device.state.mode,
-			fanSpeed: device.state.fanSpeed,
-			swing: device.state.verticalSwing,
-			horizontalSwing: device.state.horizontalSwing,
-			light: device.state.light
-		}
+		// If nothing has changed should we skip...? Like we do in StateHandler for SET?
 
-		smartModeState.lowTemperatureThreshold = device.state.targetTemperature - (device.usesFahrenheit ? 1.8 : 1)
-		smartModeState.lowTemperatureState = {
-			on: true,
-			targetTemperature: device.state.targetTemperature,
-			temperatureUnit: device.temperatureUnit,
-			mode: device.state.mode,
-			fanSpeed: device.state.fanSpeed,
-			swing: device.state.verticalSwing,
-			horizontalSwing: device.state.horizontalSwing,
-			light: device.state.light
+		if (!(device.state instanceof Classes.InternalAcState)) {
+			// TODO: log warning
+			return
 		}
-	}
 
-	// StateHandler is invoked as a Proxy, and therefore overwrites/intercepts the default get()/set() commands [traps]
-	// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy
+		const smartModeState = device.state.smartMode
+
+		smartModeState.type = 'temperature'
+		smartModeState.highTemperatureWebhook = null
+		smartModeState.lowTemperatureWebhook = null
+
+		if (device.state.mode === 'COOL') {
+			smartModeState.highTemperatureThreshold = device.state.targetTemperature + (device.usesFahrenheit ? 1.8 : 1)*climateReactAutoSetupMultiplier + climateReactAutoSetupOffset
+			smartModeState.highTemperatureState = {
+				on: true,
+				targetTemperature: device.state.targetTemperature,
+				temperatureUnit: device.temperatureUnit,
+				mode: device.state.mode,
+				fanSpeed: device.state.fanSpeed,
+				swing: device.state.verticalSwing,
+				horizontalSwing: device.state.horizontalSwing,
+				light: device.state.light ? 'on' : 'off'
+			}
 
-	// NOTE: device.state is of "type" StateHandler. When one of its properties is "set" (e.g. device.state.<property> = <val>),
-	//       that's where we actually send commands to the appropriate Sensibo devices. If a property is not set, the aformentioned
-	//       code will not execute and the changes would not take effect.
-	//
-	//       For example, if we set a property of smartMode directly, e.g. device.state.smartMode.enabled = true, StateHandler's
-	//       setter will not get called and so any changes will not take effect. This is why we MUST update a device's property as
-	//       a whole, and do it only once (otherwise's the setter will get called multiple times which will send repeated commands
-	//       to the Sensibo devices).
-	device.state.smartMode = smartModeState
-}
+			smartModeState.lowTemperatureThreshold = device.state.targetTemperature - (device.usesFahrenheit ? 1.8 : 1)*climateReactAutoSetupMultiplier + climateReactAutoSetupOffset
```

Comment:
Nope, this is fine. I'll explain the rational - the point of the offset is to allow something like this:
For example, when setting cool 25C, instead of keeping the temperature between 24 and 26, with an offset of 1, you can have it kept between 25 and 27. I actually use that for that exact purpose because I rather the AC stop cooling immediately when it reaches below 25 but also don't want it to start it at 26 because then it starts and stops every 5 minutes which drives me crazy. I rather have it start at 27 because then it only starts and stops every 10-15 minutes or so.